src/Compilers/CSharp/Test/Emit2/Attributes/AttributeTests.cs

```diff
@@ -10546,6 +10552,101 @@ class C3 { }
                 Diagnostic(ErrorCode.ERR_BadTypeArgument, "TypedReference").WithArguments("System.TypedReference").WithLocation(11, 7));
         }
 
+        [Fact]
+        public void GenericAttributeRestrictedTypeArgument_B()
+        {
+            var source = @"
+using System;
+class Attr<T> : Attribute { }
+
+unsafe class Outer
+{
+    [Attr<int*>] // 1
+    class C1 { }
+
+    [Attr<delegate*<int, void>>] // 2
+    class C2 { }
+
+    [Attr<TypedReference>] // 3
+    class C3 { }
+}
+";
+            var comp = CreateCompilation(source, options: TestOptions.UnsafeDebugDll);
+            comp.VerifyDiagnostics(
+                // (7,11): error CS0306: The type 'int*' may not be used as a type argument
+                //     [Attr<int*>] // 1
+                Diagnostic(ErrorCode.ERR_BadTypeArgument, "int*").WithArguments("int*").WithLocation(7, 11),
+                // (10,11): error CS0306: The type 'delegate*<int, void>' may not be used as a type argument
+                //     [Attr<delegate*<int, void>>] // 2
+                Diagnostic(ErrorCode.ERR_BadTypeArgument, "delegate*<int, void>").WithArguments("delegate*<int, void>").WithLocation(10, 11),
+                // (13,11): error CS0306: The type 'TypedReference' may not be used as a type argument
+                //     [Attr<TypedReference>] // 3
+                Diagnostic(ErrorCode.ERR_BadTypeArgument, "TypedReference").WithArguments("System.TypedReference").WithLocation(13, 11));
+        }
+
+        [Fact]
+        public void GenericAttributePointerArray_1()
+        {
+            var source = @"
+using System;
+class Attr<T> : Attribute { }
+
+[Attr<int*[]>] // 1
+class C1 { }
+";
+            var comp = CreateCompilation(source, options: TestOptions.UnsafeDebugDll);
+            comp.VerifyDiagnostics(
+                // (5,7): error CS0214: Pointers and fixed size buffers may only be used in an unsafe context
+                // [Attr<int*[]>] // 1
+                Diagnostic(ErrorCode.ERR_UnsafeNeeded, "int*").WithLocation(5, 7));
+        }
+
+        [Fact]
+        public void GenericAttributePointerArray_CSharp11()
+        {
+            // Legal in C#11.  Allowed for back compat
+            var source = @"
+using System;
+class Attr<T> : Attribute { }
+
+[Attr<int*[]>] // 1
+class C1 { }
+";
+            var comp = CreateCompilation(source, options: TestOptions.UnsafeDebugDll, parseOptions: TestOptions.Regular11);
+            comp.VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void GenericAttributePointerArray_CSharp12_A()
+        {
+            var source = @"
+using System;
+class Attr<T> : Attribute { }
+
+[Attr<int*[]>] // 1
+unsafe class C1 { }
+";
+            var comp = CreateCompilation(source, options: TestOptions.UnsafeDebugDll, parseOptions: TestOptions.RegularPreview);
```

Comment:
>RegularPreview [](http://example.com/codeflow?start=112&length=14)

If you're trying to test C# 12 explicitly, then RegularNext is the correct option (before C# 12 ships, it will be replaced with `Regular12` and stay that way).
Same comment applies below.

Reply:
Is there a reason for regularpreview?

Reply:
Yes, it's used by most tests that aren't specifically about a fixed language version. Most tests use it and simply keep rolling forward

Reply:
Hrmm. What's the difference between that, and not specifying a version?

Reply:
Not specifying a version is the same as RegularPreview

Reply:
Ok... so why do the tests use regularpreview then :)  why not just not supply anything?  that would be simpler/clearer IMO.  now there are two forms of tests which seem identical but are written a different way.  e.g. `CreateCompilation(source)` and `CreateCompilation(source, parseOptions: RegularPreview)`.  It would be good to standardize on one way to do this for uniformity.

Reply:
Would it be ok to go back and change all the existin gtests to not do this?  This goes back to the feedback from before that hte inconsistency is def problematic when coming in and making changes (i.e. there's consistency that is expected to be followed, and existing code that should not be followed.  determining which case you are in is unclear).  Work to make the codebase consistent would make contributions much simpler and clearer.

Reply:
That'd effectively be a style-only change, which we tend to resist. On the other hand, it's only in tests, so I think that'd be okay.

Reply:
> Would it be ok to go back and change all the existin gtests to not do this?

I do not think this would be desired.

Reply:
> It would be good to standardize on one way to do this for uniformity.

We don't have such standard and, I personally, will not support introducing such a standard. It is up to each developer to decide what values to supply explicitly. We value diversity.